lib/util/aws/cloudfront_signed_url.ex

```diff
@@ -48,8 +77,47 @@ defmodule Antikythera.Aws.CloudfrontSignedUrl do
     ]
   end
 
-  defunp create_signature(policy_statement :: v[String.t], private_key :: v[String.t]) :: String.t do
-    :public_key.sign(policy_statement, :sha, decode_rsa_key(private_key))
+  defunpt generate_ip_address_string(addresses :: v[[String.t]]) :: v[String.t] do
+    case addresses do
+      [address]     -> ~s/"#{String.trim(address)}"/
+      [address | t] -> Enum.reduce(t, ~s/["#{String.trim(address)}"/, fn (address, result) -> result <> ~s/,"#{String.trim(address)}"/ end) <> "]"
+    end
+  end
+
+  defunpt generate_custom_policy(encoded_url :: v[String.t], expires_in_seconds :: v[pos_integer], optional_policy :: Keyword.t) :: v[String.t] do
+    date_greater_than =
+      case Keyword.fetch(optional_policy, :date_greater_than) do
+        {:ok, t} when is_integer(t) and t >= 0 -> ~s/,"DateGreaterThan":{"AWS:EpochTime":#{t}}/
+        _                                      -> ""
+      end
+    ip_address =
+      case Keyword.fetch(optional_policy, :ip_address) do
+        {:ok, []}        -> ""
+        {:ok, addresses} ->
+          # It is important to restrict a connection using IpAddress, so this raises a MatchError instead of just ignoring :ip_address.
+          true = is_list(addresses)
+          ~s/,"IpAddress":{"AWS:SourceIp":#{generate_ip_address_string(addresses)}}/
```

Comment:
Maybe we can use `when`
```suggestion
        {:ok, addresses} when is_list(addresses) -> ~s/,"IpAddress":{"AWS:SourceIp":#{generate_ip_address_string(addresses)}}/          
```